### PR TITLE
[runtime/NodeJS] Instrument with OpenTelemetry

### PR DESCRIPTION
**Pull request by RobbieMcKinstry**
# Description

This PR adds telemetry information to the NodeJS runtime.
Right now, the NodeJS runtime uses whatever tracing URL the language host provides as the Zipkin destination (that might need to change in the future). 

Fixes https://github.com/pulumi/pulumi/issues/10180

### Remaining TODOs

- [x] Confirm traces are correctly sent to the right destination and shutdown is handled correctly.
- [x] Transform Pulumi engine's URL to match format expected by the OpenTelemetry Zipkin exporter.
- [x] Instrument a greater number of function calls.

### Known Issues

- [ ] This PR uses a SimpleSpanProcessor, which uploads spans as it goes. A BatchSpanProcessor is preferred, but the shutdown logic isn't working. Will open a separate issue to address that.

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works

- [ ] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

File: sdk/nodejs/cmd/run/run.ts [hunk 1]
```diff
@@ -187,7 +200,9 @@ export function run(
     const tsConfigPath: string = process.env["PULUMI_NODEJS_TSCONFIG_PATH"] ?? defaultTsConfigPath;
     const skipProject = !fs.existsSync(tsConfigPath);
 
+    span.setAttribute("typescript-enabled", false);
```
Comment (t0yv0) [inline sdk/nodejs/cmd/run/run.ts:203]: `span.setAttribute("typescript-enabled", typeScript);` perhaps.

File: sdk/nodejs/cmd/run/run.ts [hunk 2]
```diff
@@ -169,6 +171,17 @@ export function run(
...
+
+    // Start tracing. Before exiting, gracefully shutdown tracing, exporting 
+    // all remaining spans in the batch.
+    const tracingUrl: string | boolean = argv["tracing"];
+    if(typeof tracingUrl === "string" && tracingUrl.length > 0) {
+        tracing.start(tracingUrl);
+        process.on("exit", tracing.stop);
+    }
+    // Start a new span, which we shutdown at the bottom of this method.
+    const span = tracing.newSpan('language-runtime.run');
```
Comment (t0yv0) [inline sdk/nodejs/cmd/run/run.ts:183]: Does the span need closing?

File: sdk/nodejs/cmd/run/run.ts [hunk 3]
```diff
@@ -22,8 +22,10 @@ import * as ini from "ini";
 import * as semver from "semver";
 import { ResourceError, RunError } from "../../errors";
 import * as log from "../../log";
+import * as opentelemetry from "@opentelemetry/api";
```
Comment (t0yv0) [inline sdk/nodejs/cmd/run/run.ts:25]: Interesting, curious if this works. The wire protocol that Pulumi will accept is Zipkin V1 traces. Does opentelemetry clients support that, e.g. do you see the traces? I now the protocol does not support full gamut of features like reporting events and metrics.

**Comment by t0yv0** [comment]
I've got a few nits but this is looking good! Especially if we have evidence it works!

See `func TestTracePropagationGo(t *testing.T)` if you are looking for prior art of adding integration tests for this.

File: sdk/nodejs/package.json [hunk 4]
```diff
@@ -22,7 +29,7 @@
         "semver": "^6.1.0",
         "source-map-support": "^0.5.6",
         "ts-node": "^7.0.1",
-        "typescript": "~3.7.3",
+        "typescript": "~3.8.3",
```
[the remark this answers is not in the document]
Comment (RobbieMcKinstry) [reply sdk/nodejs/package.json:32]: Happy to lift it to another PR and changelog it, I was thinking the same thing.

It's required, unfortunately. OpenTelemetry uses the `export type` syntax we discussed yesterday, which is only supported In v3.8.

Personally, I don't think the change will be a problem for users, but what do I know!

**Comment by RobbieMcKinstry** [comment]
> Especially if we have evidence it works!
I've had intermittent success with it working. Not so consistent! I'm trying to determine if it's an error in my configuration or if the program is exiting before flushing the traces...

**Comment by t0yv0** [comment]
> Handle case where file URL is provided instead of Zipkin HTTP URL.

This should never be provided to nodejs code.

**Comment by RobbieMcKinstry** [comment]
@t0yv0 Yep, you're right, thanks for clarifying in my standup post yesterday.

The URL provided by the engine is still a little different than the URL I'm having success with locally (e.g. Zipkin v1 vs v2 APIs, HTTP vs TCP protocol). Doing a little more experimenting to see if they're interoperable.

File: sdk/nodejs/package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (t0yv0) [reply sdk/nodejs/package.json:32]: I'm a bit hesitant too so I think let's do that extra PR to be sure.

**Comment by t0yv0** [comment]
Ping me when ready to review. Excited to land this!

**Comment by AaronFriel** [comment]
Nit: Changelog & PR should say OpenTelemetry.

File: sdk/nodejs/package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (RobbieMcKinstry) [reply sdk/nodejs/package.json:32]: https://github.com/pulumi/pulumi/pull/10618

File: sdk/nodejs/cmd/run/run.ts (hunk 3 above)
Comment (RobbieMcKinstry) [reply sdk/nodejs/cmd/run/run.ts:25]: I was able to see the traces when running the Jaegar server, and sending it to the Zipkin endpoint using the v1 API. Not seeing the same right now with the Pulumi Zipkin endpoint.

**Comment by RobbieMcKinstry** [comment]
Update: Took another pass at this, using the Zipkin URL provided by the engine and transforming it to work with the OpenTelemetry Zipkin URL. Unfortunately, I'm now not getting the traces sent though! This seems to indicate that the Pulumi Zipkin server isn't implementing the OpenZipkin standard as strictly as the Jaeger Docker image is implementing the standard.

More tomorrow!

**Comment by RobbieMcKinstry** [comment]
One more update: Posting spans to the Zipkin v1 API fails to label the `serviceName` field. So we can collect spans from NodeJS, but it shows up unlabeled. There does not appear to be a workaround for this.

Conversely, we can send spans to a v2 API, except the rest of the engine, the host, etc doesn't support V2, so _**only**_ the NodeJS runtime's spans will show up.

Very sad.

**Comment by t0yv0** [comment]
Am I summarizing correctly?

- in the case of `--tracing file:./up.trace`, Pulumi stands up an AppDash-specific tcp proxy server and sends that URL to the nodejs runtime; nodejs does not know how to trace to that nor do we want to support AppDash long-term so this is not working

- in the case of `--tracing <zipkin-v1-endpoint>`, nodejs runtime receives that as-is but does not know how to send v1 protocol traces, only v2, so we're not going to receive useful traces from it

The right call is parking this to blocked on https://github.com/pulumi/pulumi/issues/10545 I think.

**Comment by RobbieMcKinstry** [comment]
> Am I summarizing correctly?

Not quite, though close:

> in the case of --tracing <zipkin-v1-endpoint>, nodejs runtime receives that as-is but does not know how to send v1 protocol traces, only v2, so we're not going to receive useful traces from it

NodeJS doesn't sends a v2 protocol trace to the v1 endpoint, which is received and stored by the v1 API server. However, the "name" of the span is _**not**_ sent correctly due to incompatibilities between v1 and v2. As a result, the span shows up in the UI and is inspectable, but it shows up with an unknown name.

> The right call is parking this to blocked on https://github.com/pulumi/pulumi/issues/10545 I think.

The tracing information is still useful, IMO I'd prefer putting NodeJS tracing behind the `PULUMI_EXPERIMENETAL` flag and merging, but if you still would rather park this for now, that's okay too.

**Comment by t0yv0** [comment]
OK that sounds good then. Can you indicate when ready for review, let's get this in then.

**Comment by RobbieMcKinstry** [comment]
@t0yv0 Ready for review!

**Comment by RobbieMcKinstry** [comment]
Waiting on tests, but otherwise ready for review @t0yv0

File: sdk/nodejs/cmd/run/tracing.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+// serviceName is the name of this service in the Pulumi
+// distributed system, and the name of the tracer we're using.
+const serviceName = "nodejs-runtime";
+
+// If the URL provided matches the default generated by the engine,
+// we must transform it to be compatible with the OpenTelemetry Zipkin
+// library.
+function validateUrl(destination: string): boolean {
+  if (destination.startsWith("tcp://127.0.0.1")) {
```
Comment (t0yv0) [inline sdk/nodejs/cmd/run/tracing.ts:38]: This could perhaps be just `tcp://`

File: sdk/nodejs/cmd/run/run.ts [hunk 6]
```diff
@@ -22,8 +22,10 @@ import * as ini from "ini";
...
 import * as stack from "../../runtime/stack";
 import * as settings from "../../runtime/settings";
+import * as tracing from "./tracing";
```
Comment (t0yv0) [inline sdk/nodejs/cmd/run/run.ts:28]: Consider lazy-load here to not pull any more dependencies than is needed especially when we're not tracing? Perhaps this is a small library but it seems there's a lot of transitive dependencies via opentelemetry code. Not sure how large they are though and how many files.

**Review by t0yv0 — APPROVED** [review]
LGTM!

Great if you see the finalizers actually running. 

I had an issue in the Go language provider that even on happy path Pulumi CLI would send a SIGKILL to the pulumi-language-go process as well as any child processes it detects, which also currently applies to pulumi-language-nodejs process.

Perhaps this does not apply here.

Finalizers on the `main()` scope of the language provider cannot be relied on. But finalizers on the scope of an individual gRPC method such as `Run()` can be relied on because the Pulumi CLI will await the end of a gRPC call. Perhaps that's the case here.

File: sdk/nodejs/cmd/run/run.ts (hunk 6 above)
Comment (RobbieMcKinstry) [reply sdk/nodejs/cmd/run/run.ts:28]: I'm not sure its entirely avoidable: other files throughout the runtime inevitably call `getTracer().newSpan()` regardless of whether OpenTelemetry is enabled or not. That's the API pattern OTel expects -- everything is a no-op unless initialized, so client code should look the same whether or not tracing is occurring.

We could add a shim over the OTel API to create our own no-import, no-op trace API but we'd probably sacrifice a lot of conveniences of the OTel API. Happy to discuss further if you're unsatisfied with my justification :)

**Comment by AaronFriel** [comment]
bors retry

**Pull request merged**